README
ecmh - Easy Cast du Multi Hub

Easy Cast du Multi Hub (ecmh) is a networking daemon that relays
and aggregates multicast packets from/to network interfaces.

This allows IPv6 multicast routing on Linux and other OS's that
do not implement IPv6 multicast routing. It also allows IPv4 to
IPv6 and IPv6 to IPv4 translation of multicast traffic.
Allowing multicast where it is not available at the moment.
It also has a tunnel mode so that it can detect multicast
packets being tunneled inside protocol 41 (IPv6 in IPv4)
tunnels (RFC3056). This is for instance very handy for
projects like SixXS (http://www.sixxs.net) where we can
install this daemon on the POPs and let it easily do
multicast IPv6 allowing the m6bone (http://www.m6bone.net)
to grow and provide IPv6 multicast and because of the
translation also IPv4 multicast everywhere we would want it to be.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

To contact the author send a email to:

Jeroen Massar <jeroen@unfix.org>

Authors Website: http://unfix.org/~jeroen/
   ecmh Website: http://unfix.org/projects/ecmh/



doc/changelog
ecmh (2013.09.28-1) stable; urgency=low

  * Update License to BSD 3-clause
  * Update email address

 -- Jeroen Massar <jeroen@massar.ch>  Sat, 28 Sep 2013 01:20:42 +0100

ecmh (2005.02.09-1) stable; urgency=medium

  * Fixed MLDv2, should now work correctly.
  * Tested against live Cisco connected to m6bone
    Thanks to: Bernhard Schmidt & Marco d'Itri
  * Upstream interface selection (-i)
      No queries will be sent to that interface.
  * Very Strict compilation warnings.

 -- Jeroen Massar <jeroen@massar.ch>  Wed, 09 Feb 2005 18:55:42 +0100

ecmh (2004.11.24-1) stable; urgency=low

  * Releasing as 2004.11.24
  * MLDv2 Reporting
  * Interface lookups now in an array for faster lookup
  * Promisc mode option, which is useful for broken cards
  * Tigher mld[1|2]only modes
  * Not specifying -DECMH_SUPPORT_MLD2 completely disabled MLDv2 code
  * -DECMH_GETIFADDR option for turning on/off the usage of getifaddr()
    useful when the kernel does not support it correctly
  * Higher verbosity

 -- Jeroen Massar <jeroen@massar.ch>  Wed, 24 Nov 2004 23:52:42 +0100

ecmh (2004.10.10-1) stable; urgency=low

  * Releasing as 2004.10.10
  * FreeBSD 4 support
  * Less verbosity
  * Detection of IPv4 update changes

 -- Jeroen Massar <jeroen@massar.ch>  Sun, 10 Oct 2004 13:12:42 +0200

ecmh (2004.10.09-1) stable; urgency=low

  * Releasing as 2004.10.09.
  * Multiple IPv4 aliases on one interface and then using an alias
    for tunneling
  * Ignoring of 6to4 anycast address when wanting to do proto-41
  * Only update the interface list every 5 minutes avoiding
    refreshes when there is nothing to refresh or when it
    just has been done.

 -- Jeroen Massar <jeroen@massar.ch>  Sat, 9 Oct 2004 00:00:42 +0200

ecmh (2004.10.08-1) stable; urgency=low

  * Releasing as 2004.10.08.
  * Interface destroy fix

 -- Jeroen Massar <jeroen@massar.ch>  Fri, 8 Oct 2004 16:15:42 +0200

ecmh (2004.10.07-1) stable; urgency=low

  * Releasing as 2004.10.07.
  * FreeBSD support.
  * Special tunneling mode.
  * MLDv2 implemented and enabled.

 -- Jeroen Massar <jeroen@massar.ch>  Thu, 7 Oct 2004 11:23:42 +0200

ecmh (2004.02.18-1) stable; urgency=low

  * Releasing as 2004.02.18.
  * Added TTL decrementing, streams must set their Hop Limit
    to something like the usual 64 hops otherwise they will
    be discarded, as it is multicast, we do not return errors.
  * ICMPv6 Echo Requests work and nicely reply.
    Traceroutes should work using this btw.
    ping6 -I eth0 -t 1 radio.unfix.org, shows the first hop.
    ping6 -I eth0 -t 1 radio.unfix.org, shows the second one, etc.
  * Timeout now handled outside of signal handler, avoids
  * Gathers one global address per interface, this one is
    used when replying to ICMPv6 echo requests.
  * Added Hop Limit Exceeded counter.
  * Changed timeouts to the ECMH_ROBUSTNESS_FACTOR, it is now
    set to 3 instead of 2.
  * Added ECMH_OPTIONS to the main Makefile.
  * Noted in the src/Makefile that one should build from the main tree.
  * Added a help target to the toplevel Makefile.

 -- Jeroen Massar <jeroen@massar.ch>  Tue, 17 Feb 2004 19:46:00 +0100

ecmh (2004.02.17-1) unstable; urgency=low

  * Releasing as 2004.02.17.
  * More debugging.
  * Hopefully finally fixed lists and thus timeouts.
  * Using LIST_LOOP2() when deleting inside a loop.
  * Reductions to 0 cause an immediate MLD Query to be
    sent for only that specific group.
  * inet_ntop() argument fixes and using INET[6]_ADDRSTRLEN now.
  * -f argument fix, correctly setting it to false instead of true.
  * Set ECMH_SUBSCRIPTION_TIMEOUT to 125 per MLDv2 section 9.2.
  * Using list_delete_all_node(). instead of list_delete().

 -- Jeroen Massar <jeroen@massar.ch>  Mon, 17 Feb 2004 01:18:00 +0100

ecmh (2004.01.16-1) unstable; urgency=low

  * Fixed timing issues
  * Complete MLDv2 disabling
  * Source code cleanup
  * Added more beta MLDv2 functionality

 -- Jeroen Massar <jeroen@massar.ch>  Sun, 16 Feb 2004 02:41:00 +0100

ecmh (2004.01.11-1) unstable; urgency=low

  * Comparing argv[2] != "-f" (Reported by Guus Sliepen)
    now using getopt_long also allowing future options.
  * Drop root privs, use -u to specify to which user
    Defaults to 'nobody' for debian.
  * Now also filters non-multicast from reports/leaves
  * Logs destination IPv6 vs MLD report MCA differences
  * Changed COMPARE_IPV6_ADDRESS into IN6_ARE_ADDR_EQUAL statements
  * Fixed starttime in dump output
  * Added subscription ages in stats output
  * Stats file is now opened
  * Shutdown is now showed in the log
  * Added cleanup at the end

 -- Jeroen Massar <jeroen@massar.ch>  Sun, 11 Jan 2004 22:02:42 +0100

ecmh (2004.01.10-1) unstable; urgency=low

  * Initial package of the initial version.
  * This package will be deployed on the SixXS POPs running Debian
  * MLDv2 support is DISABLED as that has not been tested properly.
  * Checked into SourceForge CVS
  * Uploaded to SourceForge

 -- Jeroen Massar <jeroen@massar.ch>  Sat, 10 Jan 2004 20:00:42 +0100



src/ecmh.h
/**************************************
 ecmh - Easy Cast du Multi Hub
 by Jeroen Massar <jeroen@massar.ch>
**************************************/

#ifndef ECMH_H
#define ECMH_H "ECMH"

#ifdef __GNUC__
#define PACKED __attribute__((packed))
#define ALIGNED __attribute__((aligned))
#define UNUSED __attribute__ ((__unused__))
#else
#define PACKED
#define ALIGNED
#define UNUSED
#endif

#include <stdint.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <strings.h>
#include <ctype.h>
#include <time.h>
#include <signal.h>
#include <syslog.h>
#include <pwd.h>
#include <getopt.h>
#include <assert.h>
#include <inttypes.h>

#include <net/if.h>
#include <netinet/if_ether.h>
#include <sched.h>
#ifdef linux
#include <netpacket/packet.h>
#endif
#if defined(__FreeBSD__) || defined(__MACH__)
#include <fcntl.h>
#include <sys/uio.h>
#include <netinet/in_systm.h>
#include <net/ethernet.h>
#include <net/bpf.h>
#define ETH_P_IPV6 ETHERTYPE_IPV6
#define ETH_P_IP ETHERTYPE_IP
#endif
#include <ifaddrs.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <netinet/icmp6.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>
#include <sys/ioctl.h>

#ifdef __MACH__
#include <mach/clock.h>
#include <mach/mach.h>
#endif

#include "mld.h"

#define PIDFILE "/var/run/ecmh.pid"
#define ECMH_DUMPFILE "/var/run/ecmh.dump"

#define ECMH_VERSION_STRING "Easy Cast du Multi Hub (ecmh) %s/%s by Jeroen Massar <jeroen@massar.ch>\n"

#ifdef DEBUG
#define D(x) x
#else
#define D(x) {}
#endif

/* The timeout for queries */
/* as per RFC3810 MLDv2 "9.2.  Query Interval" */
#define ECMH_SUBSCRIPTION_TIMEOUT	125

/* Robustness Factor, per RFC3810 MLDv2 "9.1.  Robustness Variable" */
#define ECMH_ROBUSTNESS_FACTOR		2

#ifndef ICMP6_MEMBERSHIP_QUERY
#define ICMP6_MEMBERSHIP_QUERY	MLD_LISTENER_QUERY
#endif

#ifndef ICMP6_MEMBERSHIP_REPORT
#define ICMP6_MEMBERSHIP_REPORT MLD_LISTENER_REPORT
#endif

#ifndef ICMP6_MEMBERSHIP_REDUCTION
#define ICMP6_MEMBERSHIP_REDUCTION MLD_LISTENER_REDUCTION
#endif

#include "linklist.h"
#include "common.h"

/* Booleans */
#define false	0
#define true	(!false)
#define bool	uint64_t

#include "interfaces.h"
#include "groups.h"
#include "grpint.h"
#include "subscr.h"

/* Our configuration structure */
struct conf
{
	uint64_t		maxgroups;
	uint64_t		maxinterfaces;			/* The max number of interfaces the array can hold */
	struct intnode		*ints;				/* The interfaces we are watching */
	struct list		*groups;			/* The groups we are joined to */

	char			*upstream;			/* Upstream interface */
	uint64_t		upstream_id;			/* Interface ID of upstream interface */

	bool			daemonize;			/* To Daemonize or to not to Daemonize */
	bool			verbose;			/* Verbose Operation ? */
	bool			quit;				/* Global Quit signal */
#ifdef ECMH_SUPPORT_MLD2
	bool			mld1only;			/* Only MLDv1 ? */
	bool			mld2only;			/* Only MLDv2 ? */
#endif
	bool			promisc;			/* Make interfaces promisc? (To be sure to receive all MLD's) */
	
	uint8_t			*buffer;			/* Our buffer */
	uint64_t		bufferlen;			/* Length of the buffer */

#ifndef ECMH_BPF
	int			rawsocket;			/* Single RAW socket for sending and receiving everything */
	int			__padding;
#else
	bool			tunnelmode;			/* Intercept&handle proto-41 packets? */
	struct list		*locals;			/* Local devices that could have tunnels */
	fd_set			selectset;			/* Selectset */
	uint64_t		hifd;				/* Highest File Descriptor */
#endif

	FILE			*stat_file;			/* The file handle of ourdump file */
	time_t			stat_starttime;			/* When did we start */
	uint64_t		stat_packets_received;		/* Number of packets received */
	uint64_t		stat_packets_sent;		/* Number of packets forwarded */
	uint64_t		stat_bytes_received;		/* Number of bytes received */
	uint64_t		stat_bytes_sent;		/* Number of bytes forwarded */
	uint64_t		stat_icmp_received;		/* Number of ICMP's received */
	uint64_t		stat_icmp_sent;			/* Number of ICMP's sent */
	uint64_t		stat_hlim_exceeded;		/* Packets that where dropped due to hlim == 0 */
};

#ifndef ETH_P_IPV6
#define ETH_P_IPV6	0x86dd
#endif

/* Global Stuff */
extern struct conf *g_conf;

#endif /* ECMH_H */
